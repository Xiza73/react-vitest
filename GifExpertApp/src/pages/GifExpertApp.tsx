import { Browser, CategoriesContainer } from "@/components";
import { useState } from "react";

const initialCategories = ["One Punch", "Dragon Ball"];

export const GifExpertApp: React.FC = () => {
  const [categories, setCategories] = useState<string[]>(initialCategories);
  const [browserValue, setBrowserValue] = useState<string>("");

  const handleAddCategory = () => {
    if (browserValue.trim().length === 0) return;
    if (
      categories.includes(browserValue) ||
      categories.includes(browserValue.toLowerCase())
    ) {
      alert("La categor√≠a ya existe");
      return;
    }
    setCategories([browserValue, ...categories]);
    setBrowserValue("");
  };

  return (
    <main>
      <h1>GifExpertApp</h1>
      <div></div>
      <Browser
        buttonText="Agregar"
        onChange={(e) => setBrowserValue(e.target.value)}
        onSubmit={handleAddCategory}
        value={browserValue}
        placeholder="Buscar gifs..."
      />
      <CategoriesContainer titles={categories} />
    </main>
  );
};
