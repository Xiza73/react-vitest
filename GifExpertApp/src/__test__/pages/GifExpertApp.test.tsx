import { GifExpertApp } from "@/pages";
import { fireEvent, render, screen } from "@testing-library/react";

describe("Tests on <GifExpertApp />", () => {
  const value = "Yoda";

  test("should match the snapshot", () => {
    const { container } = render(<GifExpertApp />);

    expect(container).toMatchSnapshot();
  });

  test("should show a list of categories", () => {
    render(<GifExpertApp />);

    const headers = screen.getAllByRole("heading", { level: 2 });

    expect(headers.length).toBe(2);
  });

  test("should change the browser input", () => {
    render(<GifExpertApp />);
    const input: HTMLInputElement =
      screen.getByPlaceholderText(/buscar gifs.../i);

    fireEvent.change(input, { target: { value } });

    expect(input.value).toBe(value);
  });

  test("should add a category when the form is submitted", () => {
    render(<GifExpertApp />);

    const input: HTMLInputElement =
      screen.getByPlaceholderText(/buscar gifs.../i);

    
    const button: HTMLButtonElement = screen.getByRole("button", {
      name: /agregar/i,
    });

    fireEvent.change(input, { target: { value } });

    expect(input.value).toBe(value);

    fireEvent.click(button);
    const headers = screen.getAllByRole("heading", { level: 2 });

    expect(headers.length).toBe(3);
    expect(input.value).toBe("");
  });
});
