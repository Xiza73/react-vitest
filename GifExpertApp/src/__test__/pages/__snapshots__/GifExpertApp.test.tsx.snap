// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests on <GifExpertApp /> > should match the snapshot 1`] = `
<div>
  <main>
    <h1>
      GifExpertApp
    </h1>
    <div />
    <div>
      <input
        class="undefined"
        placeholder="Buscar gifs..."
        value=""
      />
      <button>
        Agregar
      </button>
    </div>
    <ol>
      <li>
        <h2>
          One Punch
        </h2>
        <ol />
        <p>
          Loading...
        </p>
      </li>
      <li>
        <h2>
          Dragon Ball
        </h2>
        <ol />
        <p>
          Loading...
        </p>
      </li>
    </ol>
  </main>
</div>
`;
